### PR TITLE
PE-4949: feat(bdi strategy upload)

### PR DESCRIPTION
**Pull request by thiagocarvalhodev**
- implements the `UploadFileStrategy` to abstract the logic for how we upload the file using a bundle or data items
- implemented factories and their tests
- FIX ME: I noticed a problem on updating the status of the upload

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/6ra020rgoe8ro

### CODE REVIEW

**Review by karlprieb — DISMISSED** [review]

File: lib/components/upload_form.dart [hunk 1]
```diff
@@ -746,14 +747,26 @@ class _UploadFormState extends State<UploadForm> {
             }
 
             return ArDriveStandardModal(
-              title: appLocalizationsOf(context).uploadFailed,
+              width: kLargeDialogWidth,
+              title: 'Problem with Upload',
```
Comment (matibat) [inline lib/components/upload_form.dart:751]: Has to be localized.

File: lib/components/upload_form.dart [hunk 2]
```diff
@@ -746,14 +747,26 @@ class _UploadFormState extends State<UploadForm> {
...
-              ],
+              content: state.failedTasks != null
+                  ? _failedUploadList(state.failedTasks!)
+                  : null,
+              actions: state.failedTasks == null
+                  ? null
+                  : [
+                      ModalAction(
+                        action: () => Navigator.of(context).pop(false),
+                        title: 'Do Not Fix',
```
Comment (matibat) [inline lib/components/upload_form.dart:761]: Has to be localized.

File: lib/components/upload_form.dart [hunk 3]
```diff
@@ -746,14 +747,26 @@ class _UploadFormState extends State<UploadForm> {
...
+                      ),
+                      ModalAction(
+                        action: () {
+                          context.read<UploadCubit>().retryUploads();
+                        },
+                        title: 'Re-Upload',
```
Comment (matibat) [inline lib/components/upload_form.dart:767]: Has to be localized.

File: lib/components/upload_form.dart [hunk 4]
```diff
@@ -937,8 +950,10 @@ class _UploadFormState extends State<UploadForm> {
                         if (task.status == UploadStatus.inProgress ||
                             task.status == UploadStatus.complete ||
                             task.status == UploadStatus.failed) {
-                          progressText =
-                              '${filesize(((task.uploadItem!.size) * task.progress).ceil())}/${filesize(task.uploadItem!.size)}';
+                          if (task.uploadItem != null) {
+                            progressText =
+                                '${filesize(((task.uploadItem!.size) * task.progress).ceil())}/${filesize(task.uploadItem!.size)}';
```
Comment (matibat) [inline lib/components/upload_form.dart:955]: We could factor this out into its own testable and atomic function.

File: lib/components/upload_form.dart [hunk 5]
```diff
@@ -1150,6 +1165,101 @@ class _UploadFormState extends State<UploadForm> {
...
+    return ConstrainedBox(
+      constraints: const BoxConstraints(
+          maxHeight: 256 * 1.5, minWidth: kLargeDialogWidth),
+      child: Container(
+        padding: const EdgeInsets.all(8),
+        child: Column(
+          mainAxisSize: MainAxisSize.min,
+          children: [
+            Text(
+                'It seems there was a partial failure uploading the following file(s). The file(s) will show as failed in your drive. Please re-upload to fix.',
```
Comment (matibat) [inline lib/components/upload_form.dart:1178]: Has to be localized.

File: packages/ardrive_uploader/lib/src/data_bundler.dart [hunk 6]
```diff
@@ -322,6 +268,59 @@ class DataTransactionBundler implements DataBundler<TransactionResult> {
...
+  }) async {
+    SecretKey? key;
+
+    if (driveKey != null) {
+      key = await deriveFileKey(
+        driveKey,
+        metadata.id,
+        keyByteLength,
+      );
+      // onStartEncryption?.call();
```
Comment (matibat) [inline packages/ardrive_uploader/lib/src/data_bundler.dart:291]: Something to do here? If not we can remove the commented code.

File: packages/ardrive_uploader/lib/src/exceptions.dart [hunk 7]
```diff
@@ -0,0 +1,103 @@
+abstract class ArDriveUploaderExceptions {
```
Comment (matibat) [inline packages/ardrive_uploader/lib/src/exceptions.dart:1]: Nit
```suggestion
abstract class ArDriveUploaderException {
```

Comment (matibat) [reply packages/ardrive_uploader/lib/src/exceptions.dart:1]: Also, I'd say it should extend `Exception`.

File: packages/ardrive_uploader/lib/src/factories.dart [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+          metadataGenerator,
+          UploadCostEstimateCalculatorForAR(
+            arCostToUsd: ConvertArToUSD(),
+            arweaveService: arweaveService,
+            pstService: pstService,
+          ),
+          pstService,
+        );
+      default:
+        throw Exception('Invalid upload type');
```
Comment (matibat) [inline packages/ardrive_uploader/lib/src/factories.dart:38]: If `UploadType` is an enum with those two values then there's no default, so we shouldn't throw.

File: packages/ardrive_uploader/lib/src/factories.dart [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+      case UploadType.turbo:
+        return UploadFileUsingDataItemFiles(
+            streamedUploadFactory: _streamedUploadFactory);
+      case UploadType.d2n:
+        return UploadFileUsingBundleStrategy(
+          dataBundler: _dataBundlerFactory.createDataBundler(type),
+          streamedUploadFactory: _streamedUploadFactory,
+        );
+      default:
+        throw Exception('Invalid upload type');
```
Comment (matibat) [inline packages/ardrive_uploader/lib/src/factories.dart:92]: Same: no default needed.

File: packages/ardrive_uploader/lib/src/factories.dart [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+      return D2NStreamedUpload();
+    } else if (type == UploadType.turbo) {
+      return TurboStreamedUpload(
+        TurboUploadServiceImpl(
+          turboUploadUri: turboUploadUri,
+        ),
+      );
+    } else {
+      throw Exception('Invalid upload type');
+    }
```
Comment (matibat) [inline packages/ardrive_uploader/lib/src/factories.dart:117]: Consider using a switch-case with no default instead.

File: packages/ardrive_uploader/lib/src/streamed_upload.dart [hunk 11]
```diff
@@ -1,34 +1,22 @@
...
-import 'package:ardrive_uploader/src/d2n_streamed_upload.dart';
-import 'package:ardrive_uploader/src/turbo_streamed_upload.dart';
-import 'package:ardrive_uploader/src/turbo_upload_service_base.dart';
 import 'package:arweave/arweave.dart';
 
-abstract class StreamedUpload<T, R> {
-  Future<R> send(
+class StreamedUploadResult {
+  final bool success;
+  final Object? error;
```
Comment (matibat) [inline packages/ardrive_uploader/lib/src/streamed_upload.dart:6]: Should it be something like...
```dart
  final Exception? exception;
```

**Review by matibat — APPROVED** [review]
Approved with comments.

**Comment by thiagocarvalhodev** [comment]
@matibat's comments will be addressed on a new task.

**Pull request merged**